### ISSUE TITLE
DST causes meetings from organiser in different timezone to sync to Google with hour offset

### ISSUE THREAD

**Issue by phw198** [phw198/OutlookGoogleCalendarSync#1728]
_Originally posted by @khobar in https://github.com/phw198/OutlookGoogleCalendarSync/issues/1701#issuecomment-1790974371_

----

I'm from Poland which switched DST this week and US will be having it next week 
Google calendar is configured my local timezone which is GMT+01:00
Both old ( pre DST ) and newly created event are being synced and created with that timezone .
If they were created within my timezone ( for example my event , or colleague of mine from Poland as well ) , they were properly updated to new time. The ones outside ( from US or India ) are shifted  and are added in GMT+1  but 1h+

Let me know if you would like to create some new issue. I can atach logs but would have to find a way how to obfuscate meeting names 

Chunk from logs: 

```
2023-11-02 16:23:16,539 INFO  1 OutlookGoogleCalendarSync.Settings [400] -  ENVIRONMENT:-
2023-11-02 16:23:16,539 INFO  1 OutlookGoogleCalendarSync.Settings [401] -    Current Locale: pl-PL
2023-11-02 16:23:16,539 INFO  1 OutlookGoogleCalendarSync.Settings [402] -    Short Date Format: yyyy-MM-dd
2023-11-02 16:23:16,539 INFO  1 OutlookGoogleCalendarSync.Settings [403] -    Short Time Format: HH:mm
2023-11-02 16:23:16,552 INFO  1 OutlookGoogleCalendarSync.Settings [405] -    System Time Zone: Central European Standard Time; DST=False
...
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [255] -  CALENDAR SYNC SETTINGS
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [256] -  Profile: Default
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [258] -  Last Synced: 2023-10-30 16:39:33
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [259] -  OUTLOOK SETTINGS:-
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [259] -    Service: DefaultMailbox
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [265] -    Mailbox/FolderStore Name: 
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [266] -    Calendar: Default Calendar (ID: XXXXX)
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [267] -    Category Filter: Exclude
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [268] -    Delete When Excluded:True
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [269] -    Categories: 
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [270] -    Only Responded Invites: False
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [271] -    Filter String: g
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [273] -    GAL Blocked: True
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [274] -  GOOGLE SETTINGS:-
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [275] -    Calendar: Work EY (ID: XXXXX)
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [276] -    Exclude Declined Invites: True
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [277] -    Exclude Goals: True
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [279] -    Cloak Email: True
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [279] -  SYNC OPTIONS:-
2023-11-02 16:23:29,569 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [282] -   How
2023-11-02 16:23:29,585 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [283] -    SyncDirection: Outlook → Google
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [284] -    MergeItems: True
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [285] -    DisableDelete: False
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [286] -    ConfirmOnDelete: False
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [287] -    SetEntriesPrivate: False
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [288] -    SetEntriesAvailable: False
2023-11-02 16:23:29,586 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [288] -    SetEntriesColour: False
....
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [302] -    SingleCategoryOnly: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [302] -    Obfuscate Words: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [312] -   When
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [313] -    DaysInThePast: 1
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [314] -    DaysInTheFuture:60
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [315] -    SyncInterval: 15
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [316] -    SyncIntervalUnit: Minutes
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [317] -    Push Changes: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [318] -   What
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [319] -    AddLocation: True
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [320] -    AddDescription: False; OnlyToGoogle: True
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [321] -    AddAttendees: False <200
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [322] -    AddColours: True
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [323] -    AddReminders: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [324] -      UseGoogleDefaultReminder: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [325] -      UseOutlookDefaultReminder: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [326] -      ReminderDND: False (22:00-06:00)
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [327] -    ExcludeFree: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [328] -    ExcludeTentative: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [329] -    ExcludePrivate: False
2023-11-02 16:23:29,608 INFO  1 OutlookGoogleCalendarSync.SettingsStore.Calendar [330] -    ExcludeAllDay: False; that are marked Free: False

...
2023-11-02 16:23:33,543 INFO  14 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [1656] -  Google calendar timezone: Europe/Warsaw
2023-11-02 16:23:33,545 DEBUG  14 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [280] -  Retrieving all events from Google: 2023-11-01 -> 2024-01-02
```

**Comment by phw198** [comment]
Thanks @khobar - this at least shows your Windows system timezone and Google calendar are matching with `CET` / `Europe/Warsaw`.

Next we would need to check what timezone is being set in Google for the individual meeting from an organiser in a different country.
1. Delete the Google meeting with the offset start time
2. Increase OGCS logging to `FINE`
3. Sync O->G to recreate the meeting
4. Provide log file for the section that creates the Google item.

**Comment by khobar** [comment]
@phw198 I'm not 100% if I'm extracting proper section from logs ..Kind of afraid to upload whole of it as I'm already flying under radar with sync to google as it has potential data leak ( via meeting names etc. )  :) 

Removed meeting from google, sync O->G
Replaced actual meeting name with XXXXX in logs

```
2023-11-02 16:56:51,819 DEBUG  12 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [437] -  Processing >> 2023-09-15 15:00 (R) => "XXXXXXXXXXXX"
2023-11-02 16:56:51,819 DEBUG  12 OutlookGoogleCalendarSync.Console [294] -  2023-09-15 15:00 (R) "XXXXXXXXXXXX"
2023-11-02 16:56:51,831 DEBUG  12 OutlookGoogleCalendarSync.Recurrence [32] -  Creating Google iCalendar definition for recurring event.
2023-11-02 16:56:51,842 FINE  12 OutlookGoogleCalendarSync.Recurrence [224] -  Building RRULE
2023-11-02 16:56:51,843 FINE  12 OutlookGoogleCalendarSync.Recurrence [228] -  Determining pattern for frequency olRecursWeekly.
2023-11-02 16:56:51,845 FINE  12 OutlookGoogleCalendarSync.Recurrence [330] -  WEEKLY
2023-11-02 16:56:51,847 FINE  12 OutlookGoogleCalendarSync.Recurrence [338] -  DayOfWeekMask = 62
2023-11-02 16:56:51,848 FINE  12 OutlookGoogleCalendarSync.Recurrence [330] -  MO,TU,WE,TH,FR
2023-11-02 16:56:51,848 FINE  12 OutlookGoogleCalendarSync.Recurrence [300] -  Checking end date.
2023-11-02 16:56:51,849 FINE  12 OutlookGoogleCalendarSync.Recurrence [330] -  20231231T150000Z
2023-11-02 16:56:51,855 DEBUG  12 OutlookGoogleCalendarSync.Recurrence [48] -  RRULE:FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR;UNTIL=20231231T150000Z
2023-11-02 16:56:51,865 FINE  12 OutlookGoogleCalendarSync.OutlookOgcs.OutlookNew [815] -  Timezone "(UTC+01:00) Sarajevo, Skopje, Warsaw, Zagreb" mapped to "Europe/Warsaw"
2023-11-02 16:56:51,873 FINE  12 OutlookGoogleCalendarSync.OutlookOgcs.OutlookNew [815] -  Timezone "(UTC+01:00) Sarajevo, Skopje, Warsaw, Zagreb" mapped to "Europe/Warsaw"
```

**Comment by phw198** [comment]
Regarding the radar ;-) I note you have `Obfuscate Words: False` - you might want to play with masking confidential words or phrases. Also note the latest alpha lets you mask any of subject, location, body.

Log snippet is good, but it appears to show it syncing a meeting you created? Or at least a colleague also in Poland, and these items aren't a problem anyway? We need to do the test on one of the problematic items from an organiser in a different time zone...

**Comment by khobar** [comment]
No it was meeting created by colleague in India 
Below screenshot from outlook 
![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/5140941/335d6e42-9397-4671-abad-5226577f36ee)


I also removed and did sync to event created by US 

```
2023-11-02 17:31:19,090 DEBUG  23 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [437] -  Processing >> 2023-05-08 16:30 (R) => "YYYYYYYYYYYYYYY "
2023-11-02 17:31:19,090 DEBUG  23 OutlookGoogleCalendarSync.Console [294] -  2023-05-08 16:30 (R) "YYYYYYYYY "
2023-11-02 17:31:19,163 DEBUG  23 OutlookGoogleCalendarSync.Recurrence [32] -  Creating Google iCalendar definition for recurring event.
2023-11-02 17:31:19,175 FINE  23 OutlookGoogleCalendarSync.Recurrence [224] -  Building RRULE
2023-11-02 17:31:19,175 FINE  23 OutlookGoogleCalendarSync.Recurrence [228] -  Determining pattern for frequency olRecursWeekly.
2023-11-02 17:31:19,176 FINE  23 OutlookGoogleCalendarSync.Recurrence [330] -  WEEKLY
2023-11-02 17:31:19,182 DEBUG  23 OutlookGoogleCalendarSync.Recurrence [48] -  RRULE:FREQ=WEEKLY
2023-11-02 17:31:19,188 FINE  23 OutlookGoogleCalendarSync.OutlookOgcs.OutlookNew [815] -  Timezone "(UTC+01:00) Sarajevo, Skopje, Warsaw, Zagreb" mapped to "Europe/Warsaw"
2023-11-02 17:31:19,195 FINE  23 OutlookGoogleCalendarSync.OutlookOgcs.OutlookNew [815] -  Timezone "(UTC+01:00) Sarajevo, Skopje, Warsaw, Zagreb" mapped to "Europe/Warsaw"
```
![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/5140941/0a79776b-705e-41e4-9cdf-bd139abe5354)

![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/5140941/db15f3f1-c487-4a2e-9503-6d5a7bdd80ac)

**Comment by phw198** [comment]
Well that partly explains the issue. It is picking the timezone up as UTC+1 which it's not because these organisers are still in DST and therefore an hour different (UTC+2) until next week.

And now I see the problem! `GAL Blocked: True` - your corporation has blocked programmatic access to the Global Address Book. This access is required to dig into the organiser's timezone properly - without that, it's getting it wrong. Unfortunately, there's nothing I can do about this :disappointed:

**Comment by khobar** [comment]
That would explain the US based created events. But why it's same with India ones where there is no DST ? 
Also meeting  created on Monday (by India colleague), so after DST kicked in is on proper time :)
(will verify timezones tommorow in the morning) 

But if what you say is true, I will observe what will happen on Monday, when US will have DST change as well

**Comment by phw198** [comment]
The message in Outlook that "this meeting was adjusted to your timezone" is calculated when you receive it, I believe - so Microsoft is basically not recalculating it when DST changes. If the organiser sends a new invite, that's calculated afresh and it works. It's horrid - but that's Microsoft! And that's why I have to dig into the organisers timezone details (using the GAL)...which Microsoft holds as a text string in the organisers local language! The whole topic is a :wastebasket: :fire: 

I think there's [a workaround actually posted here](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#old-events-not-observing-daylight-saving) though - looks like I came across this back when the project was on CodePlex!

**Comment by khobar** [comment]
Today DST in US . Meetings based from that country aligned themselves after removing and resyncing , The ones from India are a gonners. Asked folks to resent outlook invites after I removed them to properly sync them  :/

**Comment by phw198** [comment]
@khobar  Did you not try the workaround of making faux edit and saving the problematic items - see if that prompts Microsoft to recalculate the DST date time?

Regardless, this is all idiosyncrasies of Microsoft and Outlook, not something that OGCS can really resolve itself - plus the problem for you is aggravated by corporate policy blocking full access. For that part I'll add a note to the start of the sync for those affected:
![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/11843015/b758c17a-411e-45ce-a0db-6787e1661927)

**Comment by khobar** [comment]
Sorry for delay .
So  first tried to go and edit invitation but,  as it was sent by somebody else there were no options to do so ...
Then I removed meeting from outlook, removed it from google , and asked for fw of the meeting again, but it appeared back again in wrong time . So it must be something with When meeting was created +/- DST 

Then I experimented and saved on meeting as `ics`, deleted it , synced , and manually changed timezone of that meeting by opening it in notepad and changed  from +0530 to : 
TZOFFSETFROM:+0430
TZOFFSETTO:+0430

only after that , meeting showed in proper time on my google calendar

**Comment by phw198** [comment]
Thanks for the update @khobar - just to clarify that the cause of OGCS incorrectly getting the start time in your situation, was  because corporate policy is stopping it from accessing the organiser's time zone and therefore doesn't know if to apply DST or not.

> So first tried to go and edit invitation but, as it was sent by somebody else there were no options to do so

You should still be able to open it, make a change and then `Ctrl+S` - at which point it gives some message about this making a change the organiser will not know about, but you can still proceed with that local change.

If for other items that still doesn't work, then your workaround of manually editing the `.ics` file is probably the best solution, unfortunately!

**Comment by phw198** [comment]
Have updated the new sync message to:
> Corporate policy/anti-virus is restricting certain functionality. This may prevent proper detection of meeting organiser time zones, resulting in incorrect local start times due to DST.

**Comment by khobar** [comment]
I don't think I can change meeting ..not via Outlook or maybe I just can't see that 
![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/5140941/5a1f8ab6-0afd-4b7a-ac57-4bec0197c108)

Also my "workaround " flopped as after a while outlook must have fixed timezones or smth and after sync meeting went back to it's "wrong" time :( 

Fun part is as you described ..If I open series on the meeting invite it shows that "meeting occurs 15.00-16.00 ( 3pm-4pm )" . If I open single occurrence  it shows 2pm-3pm which is when meeting takes place . Maybe when sync is happening meetings has to be looked as indivudal calendar meetings and don't look at the recurence ? 

I compared ICS files for whole series 
```
DTSTART:20231117T130000Z
DTEND:20231117T140000Z
```

while if I export ics for single occurence of that meeting 
```
DTEND;TZID="India Standard Time":20231115T193000
DTSTART;TZID="India Standard Time":20231115T183000
```


If it would help I can send both ics files ( redacted of course :) )  , but it might be just how outlook saves those files

**Comment by phw198** [comment]
Yes, and that's the problem - your corporate policy is stopping OGCS getting at the occurrence TZID of "India Standard Time", so it can only use the information available on the series which is UTC.

I have a Save and Close button on recurring series organised by others:
![image](https://github.com/phw198/OutlookGoogleCalendarSync/assets/11843015/c8ec4527-0b3d-4f1a-93a4-2cbd8453b0f1)
...but this is why I said to just press `Ctrl+S`. There should also be a `Save` and `Save As` under the `File` menu?
